keras_cv/models/segmentation/deeplabv3plus.py

```diff
@@ -0,0 +1,319 @@
+    """
+
+    def __init__(
+        self,
+        classes,
+        backbone,
+        spatial_pyramid_pooling=None,
+        segmentation_head=None,
+        segmentation_head_activation="softmax",
+        input_shape=(None, None, 3),
+        input_tensor=None,
+        low_level_feature_layer=None,
+        weights=None,
+        **kwargs,
+    ):
+
+        if not isinstance(backbone, tf.keras.layers.Layer):
+            raise ValueError(
+                "Backbone need to be a `tf.keras.layers.Layer`, " f"received {backbone}"
+            )
+
+        if weights and not tf.io.gfile.exists(
+            parse_weights(weights, True, "deeplabv3plus")
+        ):
+            raise ValueError(
+                "The `weights` argument should be either `None` or the path to the "
+                f"weights file to be loaded. Weights file not found at location: {weights}"
+            )
+
+        inputs = utils.parse_model_inputs(input_shape, input_tensor)
+
+        if input_shape[0] is None and input_shape[1] is None:
+            input_shape = backbone.input_shape[1:]
+            inputs = layers.Input(tensor=backbone.input, shape=input_shape)
+
+        if input_shape[0] is None and input_shape[1] is None:
+            raise ValueError(
+                "Input shapes for both the backbone and DeepLabV3Plus are `None`."
+            )
+
+        x = inputs
+        high_level = backbone(x)
+
+        if low_level_feature_layer is None:
+            if "resnet" in backbone.name:
+                low_level = backbone.get_layer("v2_stack_1_block4_1_relu").output
+            else:
+                raise ValueError(
+                    "You have to specify the name of the low-level layer in the "
+                    "model used to extract low-level features."
+                )
+        else:
+            if not isinstance(backbone, tf.keras.layers.Layer):
+                raise ValueError(
+                    "Backbone need to be a `tf.keras.layers.Layer`, "
+                    f"received {backbone}"
+                )
+            low_level = backbone.get_layer(low_level_feature_layer).output
+
+        if spatial_pyramid_pooling is None:
+            spatial_pyramid_pooling = SpatialPyramidPooling(dilation_rates=[6, 12, 18])
+
+        output = spatial_pyramid_pooling(high_level)
+        output = tf.keras.layers.UpSampling2D(
+            size=(4, 4),
+            interpolation="bilinear",
+        )(output)
+
+        output = layers.Concatenate()([output, low_level])
+
+        if segmentation_head is None:
+            segmentation_head = SegmentationHead(
+                classes=classes,
+                name="segmentation_head",
+                output_scale_factor=4,
+                convs=1,
+                dropout=0.2,
+                kernel_size=1,
+            )
+
+        # Segmentation head expects a multiple-level output dictionary
```

Comment:
can we update the head to support single Tensor?

Reply:
Updated.

I have to ask again what you think about having the Semantic Segmentation Head being a public API? Both for DeepLabV3 and DeepLabV3Plus can use the same one (with a single argument difference), and it's conceivable that FCNs could use a customized form of it (customized through args) 🤔

Reply:
If we could refactor it to support all these models, I think we should make it public API.
all these models = deeplabv3 +deeplabv3plus + fcn + unet
I think there's an ongoing PR to offer unet + segmentation?

Reply:
I'm pretty sure we can. Since accounting for all of them might prove tricky, let's postpone it until all four models have been finished, merged and tested for performance.
Once they're all on stable footing, we can refactor the head to be shared?